### PR TITLE
Attempt at state

### PR DESCRIPTION
**Pull request by euri10**
totally untested, not sure if the send / receive might need some tweaks as well

### CODE REVIEW

**Comment by euri10** [comment]
well that doesn't make it, I dont get how we pass the lifespan state to the scope

**Comment by florimondmanca** [comment]
What went wrong with this attempt? Do we still see the Starlette runtime error?

File: tests/test_manager.py [hunk 1]
```diff
@@ -221,3 +220,18 @@ async def test_lifespan_not_supported(app: typing.Callable) -> None:
     with pytest.raises(LifespanNotSupported):
         async with LifespanManager(app):
             pass  # pragma: no cover
+
+
+@pytest.mark.usefixtures("concurrency")
+async def test_lifespan_state_async_cm() -> None:
+    async def app(scope: Scope, receive: Receive, send: Send) -> None:
```
Comment (florimondmanca) [inline tests/test_manager.py:227]: It seems `.state` is public and `__aenter__()` is changed to return the manager itself just so we can run this test.

Instead I think we should inspect `scope["state"]` inside the app.

**Comment by euri10** [comment]
> Refs #57
> 
> What went wrong with this attempt? Do we still see the Starlette runtime error?

nope but I dont have the state key in the scope,

**Comment by florimondmanca** [comment]
Ah, well, I think it’s because the lifespan sends a scope with state, but it doesn’t include the state in scopes for other ASGI scope types.

On other words, LifespanManager is merely a « lifespan runner » for now, but it should actually wrap the app like a middleware so that it copies the state into all ASGI requests…

**Comment by Kludex** [comment]
> I think that shouold do the job, just couldnt see fast enough how to "pass" it from the lifespan to the app

It would be a new app, so something like:

```python
async with LifespanManager(app) as wrapped_app:
    async with AsyncClient(app=wrapped_app, base_url="http://test") as client:
        yield client
```

I wanted to confirm that was what @florimondmanca meant.

Something like the following does the job:
```diff
-    async def __aenter__(self) -> None:
+    def _include_state_middleware(self, app: ASGIApp) -> ASGIApp:
+        async def state_injector(scope: Scope, receive: Receive, send: Send) -> None:
+            scope["state"] = self._state
+            await app(scope, receive, send)
+
+        return state_injector
+
+    async def __aenter__(self) -> ASGIApp:
         await self._exit_stack.__aenter__()
         await self._exit_stack.enter_async_context(
             self._concurrency_backend.run_in_background(self.run_app)
         )
         try:
             await self.startup()
+            self.app = self._include_state_middleware(self.app)
+            return self.app
         except BaseException:
             await self._exit_stack.aclose()
             raise
```

**Pull request closed without merging**